Add tutorial on creating slack messages from enforce cloudevents (#607)
chainguard-dev/edu: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/create-slack-alerts/index.md

---

> title: "Create Slack Alerts from Enforce Cloud Events"

Reviewer: so i know the other tutorial is already published, but shouldn't this title be "Create Slack Alerts from Enforce CloudEvents"? i.e., no space between cloud and events?

Author: Oh yeah, not sure how I managed to do that, will update the other one too

> title: "Create Slack Alerts from Enforce Cloud Events"

Author: ```suggestion
title: "Create Slack Alerts from Enforce CloudEvents"
```

type: "article"
description: "Creating Slack Alerts from Enforce Cloud Events"
date: 2023-04-11T15:22:20+01:00
lastmod: 2023-04-11T15:22:20+01:00
draft: false
tags: ["Enforce", "Product", "Cloudevents"]
images: []
menu:
  docs:

>     parent: "chainguard-enforce-kubernetes"

Reviewer: ```suggestion
    parent: "cloudevents"
```

Author: Will do the same for the other one too

weight: 650
toc: true
---

> _This documentation is related to Chainguard Enforce. You can request access to the product by selecting **Chainguard Enforce** on the [inquiry form](https://www.chainguard.dev/contact?utm_source=docs)._

Chainguard Enforce for Kubernetes can send Cloudevents as webhook requests that you can subscribe to and use to do things like generate alerts, create Github issues, post messages to Slack channels and more. This tutorial is meant to serve as a companion to the [Enforce Events slack notifier](https://github.com/chainguard-dev/enforce-events/tree/main/github-issue-opener) example application. It will guide you through creating and deploying a Google Cloud Run service that will send a Slack message to a specifed channel for any Enforce policy violation or admission event.

> You can also opt to run the example application elsewhere. As long as it is publicly accessible to the [Enforce CIDR network ranges](/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/network-requirements/#cidr-ranges), you will be able to receive Cloudevents and create Slack messages with the example application. If you opt to run the demo application outside of Cloud Run, you can skip to the [Subscribing to Enforce Cloudevents](#subscribing-to-enforce-cloudevents) section, provided you know the publicly accessible URL to your deployed instance of the demo.

Reviewer: ```suggestion
You can also opt to run the example application elsewhere. As long as it is publicly accessible to the [Enforce CIDR network ranges](/chainguard/chainguard-enforce/reference/network-requirements/#cidr-ranges), you will be able to receive Cloudevents and create Slack messages with the example application. If you opt to run the demo application outside of Cloud Run, you can skip to the [Subscribing to Enforce Cloudevents](#subscribing-to-enforce-cloudevents) section, provided you know the publicly accessible URL to your deployed instance of the demo.
```

And another general reminder to double check that all the URLs are relative and updated!

## Prerequisites

To follow along with this guide, it is assumed that you have a Kubernetes cluster with Enforce enabled. If you need to set this up, follow our [Getting Started](/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/chainguard-enforce-user-onboarding/) guide to create a local [kind](https://kind.sigs.k8s.io/) cluster that you can use for experimentation.

You will also need the following:

> * The [`gcloud` CLI](https://cloud.google.com/sdk/docs/install) and a Google Cloud account if you choose to run the application using Cloud Run.

Reviewer: Like last time, format these bullets consisently

  + resource "google_cloud_run_service" "slack-notifier" {
      + autogenerate_revision_name = false
      + id                         = (known after apply)
      + location                   = "<google location>"
      + name                       = "enforce-events-slack-notifier"
      + project                    = "<gcloud project ID>"
      + status                     = (known after apply)
. . .
To perform exactly these actions, run the following command to apply:
    terraform apply "enforce-events.plan"
```

Inspect the plan to ensure that the resources that Terraform will create are in line with your Google Cloud project's settings and IAM permissions. Once you are satisfied with the plan, apply it:

```shell
terraform apply enforce-events.plan
```

Terraform will run the module and you will receive output like the following:

```
google_service_account.slack-notifier: Creating...
google_secret_manager_secret.slack-url: Creating...
. . .
Outputs:

secret-command = "echo -n YOUR SLACK WEBHOOK | gcloud --project jamon-chainguard secrets versions add enforce-events-slack-url --data-file=-"
url = "https://enforce-events-slack-notifier-ntfj7ezpja-uc.a.run.app"
```

> Note down the `url`, and the secret-command.

Reviewer: ```suggestion
Note down the `url`, and the `secret-command`.
```

## Configuring Google Cloud with your Slack Webhook URL

With the output from the `terraform apply` step, you can now run the `echo` command to configure Google Cloud with your Slack webhook URL.

> Substitute your webhok URL in place of the `YOUR SLACK WEBHOOK` string and run the following command:

Reviewer: ```suggestion
Substitute your webhook URL in place of the `YOUR SLACK WEBHOOK` string and run the following command:
```

```shell
echo -n <YOUR SLACK WEBHOOK> | gcloud --project <google project> secrets versions add enforce-events-slack-url --data-file=-
```

> This configures your Google Cloud project with your URL, so that it is available to the demo application when it is invoked on Cloud Run.

Reviewer: ```suggestion
This configures your Google Cloud project with your URL so that it is available to the demo application when it is invoked on Cloud Run.
```

### Subscribing to Enforce Cloudevents

Now that the demo application is deployed, the next step is to register it as a Cloudevents receiver using Enforce.

Run the following `chainctl` command to create an Enforce webhook subscription:

```shell
chainctl events subscriptions create \
  --group <your chainguard group> \
  https://<url from terraform>
```

> You will receive output showing the ID of the subscription, and the webhook endpoint like the following:

Reviewer: ```suggestion
You will receive output showing the ID of the subscription and the webhook endpoint like the following:
```

```
                    ID                   |             SINK
 ----------------------------------------+-----------------------------
  <chainguard group ID>/b17172905d2a4a44 | https://<url from terraform>
  ```

Enforce will now send events to the application that is running on Cloud Run at the specified URL.

> ## Configuring an Enforce Policy

Reviewer: ```suggestion
## Configuring a Chainguard Enforce Policy
```

> Before you can generate Slack messages for policy violations or admission events, you will need to ensure that you have Enforce configured with a policy that you can test. Make sure you have a cluster registered with Enforce using the `chainctl clusters ls` command. If you do not, visit our [Getting Started with Enforce Guide](https://edu.chainguard.dev/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/chainguard-enforce-user-onboarding/#step-3--prepare-kubernetes-cluster) to create and register a cluster.

Reviewer: ```suggestion
Before you can generate Slack messages for policy violations or admission events, you will need to ensure that you have Chainguard Enforce configured with a policy that you can test. Make sure you have a cluster registered with Enforce using the `chainctl clusters ls` command. If you do not, visit our [Getting Started with Enforce Guide](https://edu.chainguard.dev/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/chainguard-enforce-user-onboarding/#step-3--prepare-kubernetes-cluster) to create and register a cluster.
```

Once you have a cluster enrolled with Enforce, create the following policy from the Getting Started guide:

```
cat > sample-policy.yaml <<EOF
apiVersion: policy.sigstore.dev/v1beta1
kind: ClusterImagePolicy
metadata:
  name: sample-policy
spec:
  images:
  - glob: "ghcr.io/chainguard-dev/*/*"
  - glob: "ghcr.io/chainguard-dev/*"
  - glob: "index.docker.io/*"
  - glob: "index.docker.io/*/*"
  - glob: "cgr.dev/chainguard/**"
  authorities:
  - keyless:
      url: https://fulcio.sigstore.dev
      identities:
        - issuerRegExp: ".*"
          subjectRegExp: ".*"
EOF
```

Apply it using `chainctl`:

```
chainctl policies apply -f sample-policy.yaml --group=<your chainguard group>
```

> In the next step, you'll deploy a pod that does not match the policy, and verify that it creates a Slack message.

Reviewer: ```suggestion
In the next step, you'll deploy a pod that does not match the policy and verify that it creates a Slack message.
```

## Generating an Event that Creates a Slack Message

Create a Kubernetes deployment using `nginx`:

```shell
kubectl create deployment nginx --image=nginx
```

You should receive a message from Enforce on the command line like the following, indicating the `nginx` image is missing a signature and was not admitted into your cluster:

```
error: failed to create deployment: admission webhook "enforcer.chainguard.dev" denied the request: validation failed: failed policy: sample-policy: spec.template.spec.containers[0].image
index.docker.io/library/nginx@sha256:63b44e8ddb83d5dd8020327c1f40436e37a6fffd3ef2498a6204df23be6e7e94 signature keyless validation failed for authority authority-0 for index.docker.io/library/nginx@sha256:63b44e8ddb83d5dd8020327c1f40436e37a6fffd3ef2498a6204df23be6e7e94: no signatures found for image
```

Now check the Slack channel that you configured to receive messages. You should have a corresponding message from Enforce that looks like the following:

![Screenshot of Slack message from Enforce webhook](slack-admission-message.png)

> ## Conclusion

Reviewer: ```suggestion
## Learn More
```

> Now that you've deployed and tested sending Slack messages with the demo application, you can continue to develop it to alert about other Enforce event types. See our [Chainguard Enforce Events](/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/chainguard-enforce-events/) page for a complete reference of event types that Enforce emits.

Reviewer: ```suggestion
Now that you've deployed and tested sending Slack messages with the demo application, you can continue to develop it to alert about other Enforce event types. Check out our [Chainguard Enforce Events](/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/chainguard-enforce-events/) page for a complete reference of event types that Enforce emits.
```
